File: lib/infrastructure/recovery/file_recovery_repository.dart
import 'dart:convert';
import 'dart:io';

import 'package:foglietto/domain/recovery/recovery_filename.dart';
import 'package:foglietto/domain/recovery/recovery_note.dart';
import 'package:foglietto/domain/recovery/recovery_preview.dart';
import 'package:foglietto/domain/recovery/recovery_repository.dart';
import 'package:foglietto/infrastructure/paths/sandbox_path_provider.dart';
import 'package:path/path.dart' as p;

/// Produces UTC ISO-8601 timestamps. Injectable for deterministic tests.
typedef NowUtcProvider = DateTime Function();

/// Creates a [Directory] at the given path. Injectable so tests can inject
/// a [Directory] subclass that throws on [create] to verify error propagation.
typedef DirectoryFactory = Directory Function(String path);

/// Resolves the recovery [Directory] synchronously. Injectable for [saveSync]
/// tests and for the production lifecycle host (which obtains the base dir
/// at startup via `path_provider` and stores it before the first pause event).
///
/// Required when [FileRecoveryRepository.saveSync] is called. The production
/// caller (T-03 lifecycle host) provides this after an awaited
/// `getApplicationSupportDirectory()` at app startup.
typedef SyncRecoveryDirResolver = Directory Function();

/// Concrete [RecoveryRepository] that persists recovery files as flat `.txt`
/// files under `<applicationSupportDirectory>/recovery/`.
///
/// ## Save algorithm (FR-M2-10, FR-M2-11, FR-M2-12, §5.2)
///
/// 1. Creates the recovery directory recursively (only in the save path —
///    [SandboxPathProvider.recoveryDirectory] remains composition-only so the
///    M1 no-create test stays green).
/// 2. Computes a UTC ISO-8601 fixed-width filename with colons replaced by
///    `-`, e.g. `2026-06-13T19-20-05-123Z.txt`, ensuring lexicographic order
///    equals chronological order (R-16, NFR-M2-03).
/// 3. On filename collision (`existsSync()`), appends `-1`, `-2`, … before
///    `.txt` — never overwrites (EC-M2-07).
/// 4. Writes the text as UTF-8 (FR-M2-12).
/// 5. Returns the written [File].
///
/// ## Concurrency contract (BUG-004)
///
/// All async writes are serialised on a private [_writeChain] future. Each
/// [save] enqueues its resolve+write on the tail of the chain before
/// returning, so concurrent callers never interleave the
/// [_resolveFile]+[writeAsString] sequence.  A failing save does NOT poison
/// the chain: the internal chain tail swallows the error via [catchError]
/// while the future returned to the caller still propagates it normally
/// (EC-M2-09 is preserved).
///
/// [saveSync] does NOT participate in [_writeChain] — it is self-contained
/// and runs synchronously to completion before returning.
///
/// ## Error contract
///
/// [FileSystemException] from any I/O operation propagates to the caller
/// unchanged (EC-M2-09). The caller (lifecycle host) is responsible for
/// catching and logging it.
///
/// ## Precondition
///
/// The use case ([SaveBufferToRecovery]) guarantees `text.trim().isNotEmpty`
/// before calling [save] or [saveSync]. This class does not re-validate.
class FileRecoveryRepository implements RecoveryRepository {
  FileRecoveryRepository({
    required this._pathProvider,
    NowUtcProvider? nowUtc,
    DirectoryFactory? directoryFactory,
    this._syncRecoveryDir,
  }) : _nowUtc = nowUtc ?? _defaultNow,
       _directoryFactory = directoryFactory ?? _defaultDirectoryFactory;

  final SandboxPathProvider _pathProvider;
  final NowUtcProvider _nowUtc;
  final DirectoryFactory _directoryFactory;

  /// Synchronous recovery-directory resolver for [saveSync].
  ///
  /// Must be provided when [saveSync] will be called. The production T-03
  /// lifecycle host supplies this after an awaited `getApplicationSupportDirectory`
  /// at startup. Tests inject it directly as a closure over their temp dir.
  final SyncRecoveryDirResolver? _syncRecoveryDir;

  /// Serialises concurrent async saves so that [_resolveFile]+[writeAsString]
  /// is never interleaved across two callers (BUG-004 / EC-M2-07).
  ///
  /// The chain tail only swallows errors to avoid poisoning subsequent saves;
  /// the future returned to each individual caller still surfaces its own error.
  Future<void> _writeChain = Future<void>.value();

  static DateTime _defaultNow() => DateTime.now().toUtc();
  static Directory _defaultDirectoryFactory(String path) => Directory(path);

  @override
  Future<File> save(String text) {
    // Enqueue this save on the tail of the chain.  The returned future is the
    // individual op's future — callers observe THIS op's result or error.
    // The chain tail uses catchError so a failed op does not block later saves.
    final op = _writeChain.then((_) => _doSave(text));
    // Advance the chain on the void-typed projection so the catchError handler
    // is typed correctly (Future<void> does not require a return value).  The
    // original `op` future is returned to the caller and still propagates its
    // own error normally.
    _writeChain = op.then((_) {}).catchError((_) {});
    return op;
  }

  /// Performs the actual directory creation, filename resolution, and write.
  /// Must only be called from within the serialised [_writeChain].
  Future<File> _doSave(String text) async {
    final recoveryDir = await _pathProvider.recoveryDirectory();
    final dirToCreate = _directoryFactory(recoveryDir.path);
    await dirToCreate.create(recursive: true);

    final stem = _buildStem(_nowUtc());
    final file = _resolveFile(recoveryDir, stem);

    await file.writeAsString(text, encoding: utf8);
    return file;
  }

  // --- M5 additions (list/read/delete/deleteAll/trim) ----------------------

  @override
  Future<List<RecoveryNote>> list() async {
    final recoveryDir = await _pathProvider.recoveryDirectory();
    if (!recoveryDir.existsSync()) return const [];

    final txtFiles = _listTxtFiles(recoveryDir);

    final notes = <RecoveryNote>[];
    for (final file in txtFiles) {
      final filename = p.basename(file.path);
      final savedAt = RecoveryFilename.parse(filename);
      if (savedAt == null) continue; // skip malformed — never fatal

      final head = _readHead(file);
      final preview = RecoveryPreview.truncate(head);
      notes.add(
        RecoveryNote(path: file.path, savedAt: savedAt, preview: preview),
      );
    }

    // Return newest-first (descending by savedAt parsed from filename).
    notes.sort((a, b) => b.savedAt.compareTo(a.savedAt));
    return notes;
  }

  @override
  Future<String> read(RecoveryNote note) async {
    return File(note.path).readAsString(encoding: utf8);
  }

  @override
  Future<void> delete(RecoveryNote note) async {
    final recoveryDir = await _pathProvider.recoveryDirectory();
    if (!recoveryDir.existsSync()) return;

    final file = File(note.path);
    if (file.existsSync()) {
      await file.delete();
    }
  }

  @override
  Future<void> deleteAll() async {
    final recoveryDir = await _pathProvider.recoveryDirectory();
    if (!recoveryDir.existsSync()) return;

    for (final file in _listTxtFiles(recoveryDir)) {
      await file.delete();
    }
  }

  @override
  Future<void> trim(int keep) async {
    final recoveryDir = await _pathProvider.recoveryDirectory();
    if (!recoveryDir.existsSync()) return;

    final files = _listTxtFiles(recoveryDir);
    if (files.length <= keep) return;

    // Sort lexicographically by filename (fixed-width UTC ISO-8601 names =>
    // name order == chronological order). NEVER use mtime (NFR-M5-01).
    files.sort((a, b) => p.basename(a.path).compareTo(p.basename(b.path)));

    // Delete the oldest (smallest names) — the N-keep at the front.
    final toDelete = files.sublist(0, files.length - keep);
    for (final file in toDelete) {
      await file.delete();
    }
  }

  // --- NEW synchronous save path (Defect B, C-04) --------------------------

  /// Persists [text] to the recovery store SYNCHRONOUSLY before returning.
  ///
  /// Reuses [_buildStem] and [_resolveFile] so the filename format is
  /// byte-identical to the async [save] path (C-05, NFR-M5-01). Does NOT
  /// participate in [_writeChain] — synchronous execution is inherently atomic
  /// with respect to other callers on the same isolate thread.
  ///
  /// [FileSystemException] propagates unchanged to the caller (EC-M2-08).
  ///
  /// Requires [syncRecoveryDir] to have been provided at construction time;
  /// throws [StateError] if it was not.
  @override
  File saveSync(String text, {int keep = 10}) {
    if (_syncRecoveryDir == null) {
      throw StateError(
        'FileRecoveryRepository.saveSync requires syncRecoveryDir to be '
        'provided at construction time.',
      );
    }

    final dir = _directoryFactory(_syncRecoveryDir().path);
    dir.createSync(recursive: true);

    final stem = _buildStem(_nowUtc());
    final file = _resolveFile(dir, stem);

    file.writeAsStringSync(text, encoding: utf8);

    _trimSync(dir, keep);

    return file;
  }

  /// Retains the newest [keep] `.txt` files by LEXICOGRAPHIC filename
  /// (NFR-M5-01 — NEVER mtime). No-op when file count <= [keep].
  void _trimSync(Directory dir, int keep) {
    final files = dir
        .listSync()
        .whereType<File>()
        .where((f) => p.extension(f.path) == '.txt')
        .toList();
    if (files.length <= keep) return;

    files.sort((a, b) => p.basename(a.path).compareTo(p.basename(b.path)));

    final toDelete = files.sublist(0, files.length - keep);
    for (final file in toDelete) {
      file.deleteSync();
    }
  }

  // ── Private helpers ───────────────────────────────────────────────────────

  /// Returns all `.txt` [File]s in [dir] without sorting (caller sorts).
  List<File> _listTxtFiles(Directory dir) {
    return dir
        .listSync()
        .whereType<File>()
        .where((f) => p.extension(f.path) == '.txt')
        .toList();
  }

  /// Reads at most [_previewReadLimit] bytes from the start of [file] as UTF-8.
  ///
  /// A bounded read avoids loading arbitrarily large files for a preview.
  /// [RecoveryPreview.truncate] then collapses newlines and hard-cuts to 80.
  static const int _previewReadLimit = 512;

  String _readHead(File file) {
    final raf = file.openSync();
    try {
      final bytes = raf.readSync(_previewReadLimit);
      return utf8.decode(bytes, allowMalformed: true);
    } finally {
      raf.closeSync();
    }
  }

  /// Converts [now] to the canonical filename stem, e.g. `2026-06-13T19-20-05-123Z`.
  ///
  /// Format: `YYYY-MM-DDTHH-MM-SS-mmmZ` — colons replaced by `-` so the name
  /// is filesystem-safe and lexicographic order equals chronological order.
  String _buildStem(DateTime now) {
    final y = _pad(now.year, 4);
    final mo = _pad(now.month, 2);
    final d = _pad(now.day, 2);
    final h = _pad(now.hour, 2);
    final mi = _pad(now.minute, 2);
    final s = _pad(now.second, 2);
    final ms = _pad(now.millisecond, 3);
    return '$y-$mo-${d}T$h-$mi-$s-${ms}Z';
  }

  /// Resolves the write target, appending `-1`, `-2`, … on collision.
  File _resolveFile(Directory dir, String stem) {
    var candidate = File('${dir.path}/$stem.txt');
    var suffix = 0;
    while (candidate.existsSync()) {
      suffix += 1;
      candidate = File('${dir.path}/$stem-$suffix.txt');
    }
    return candidate;
  }

  String _pad(int value, int width) => value.toString().padLeft(width, '0');
}

File: shared/src/commonMain/kotlin/com/paolosantucci/foglietto/shared/recovery/FileRecoveryRepository.kt
package com.paolosantucci.foglietto.shared.recovery

import okio.Buffer
import okio.FileSystem
import okio.IOException
import okio.Path
import okio.Path.Companion.toPath

/**
 * okio-backed [RecoveryRepository] implementation.
 *
 * Port of Dart `FileRecoveryRepository` — collapsed to ONE synchronous `save` because okio
 * I/O is blocking (the Dart `save`/`saveSync`/`_writeChain`/`callSync`/`_trimSync` machinery
 * is dropped per the assessment drop-list and EC-08/OQ-A).
 *
 * ## Ctor injection (NFR-06)
 *
 * All platform and I/O dependencies are injected:
 *   - [fileSystem]   — the okio `FileSystem` (never hardcoded as `FileSystem.SYSTEM`).
 *   - [recoveryDir]  — the base directory for recovery `.txt` files.
 *   - [now]          — returns the current [RecoveryInstant] (UTC calendar tuple);
 *                      injectable for deterministic tests.
 *
 * ## Filename stem (FR-11, NFR-05)
 *
 * `YYYY-MM-DDTHH-MM-SS-mmmZ` — pure string assembly from the 7 calendar fields; no
 * epoch-millis date math, no kotlinx-datetime. The millisecond component is zero-padded
 * to exactly 3 digits so lexicographic order equals chronological order (FR-12, R-05).
 *
 * ## Collision suffix (FR-13, EC-07)
 *
 * If the stem `.txt` already exists, appends `-1`, `-2`, … before `.txt`. Never overwrites.
 *
 * ## Directory creation (FR-16)
 *
 * `createDirectories(recoveryDir)` is called ONLY in [save]. All read/delete operations
 * no-op (empty/null result, no throw) when the directory is absent — they NEVER create it.
 *
 * ## Read tolerance (FR-18, R-06)
 *
 * `list` is guarded by `fileSystem.exists(recoveryDir)` because okio's `FileSystem.list`
 * throws on an absent directory. Per-file reads tolerate vanished/changed files. Deletes
 * use `mustExist = false`.
 *
 * ## 512-byte lenient head decode (FR-06, NFR-04)
 *
 * `list` reads at most [RecoveryPreview.PREVIEW_READ_LIMIT] bytes from the head of each
 * file via an okio `Buffer` + `readUtf8()`. This matches Dart's `allowMalformed: true`
 * and does not throw when the 512-byte window ends mid-UTF-8 sequence (U+FFFD acceptable).
 *
 * ## Trim ranking (FR-14, R-05, R-A3)
 *
 * [trim] ranks by **lexicographic filename string** (including any `-N` collision suffix),
 * never by `savedAt` / mtime. Because stems are fixed-width and lex-ordered chronologically,
 * this correctly identifies the oldest files. The `keep` parameter has NO default — the
 * literal `10` lives at the [SaveBufferToRecovery] use-case boundary.
 *
 * Spec refs: §5.1.1, §4.2, FR-11..FR-18, NFR-03, NFR-04, NFR-06; assessment R-A3, R-A4.
 */
class FileRecoveryRepository(
    private val fileSystem: FileSystem,
    private val recoveryDir: Path,
    private val now: () -> RecoveryInstant,
) : RecoveryRepository {

    // ══════════════════════════════════════════════════════════════════════
    // RecoveryRepository — write path
    // ══════════════════════════════════════════════════════════════════════

    /**
     * Persists [text] to the recovery store and returns the written file path as a [String].
     *
     * Creates [recoveryDir] recursively (only this method creates it). Builds the
     * colon-free ms stem from the injected [now] lambda. Appends `-1`/`-2`/… on
     * filename collision. I/O errors propagate UNCHANGED (FR-15, EC-05).
     */
    override fun save(text: String): String {
        fileSystem.createDirectories(recoveryDir)
        val instant = now()
        val stem = buildStem(instant)
        val targetPath = resolveFile(stem)
        fileSystem.write(targetPath) {
            writeUtf8(text)
        }
        return targetPath.toString()
    }

    // ══════════════════════════════════════════════════════════════════════
    // RecoveryRepository — trim
    // ══════════════════════════════════════════════════════════════════════

    /**
     * Retains the newest [keep] files ranked by **lexicographic filename** (NEVER mtime).
     *
     * No-op when the directory is absent or when the file count is at or below [keep].
     * [keep] has NO default — the literal `10` lives at [SaveBufferToRecovery] (FR-14, R-05).
     */
    override fun trim(keep: Int) {
        if (!fileSystem.exists(recoveryDir)) return
        val files = listTxtFiles()
        if (files.size <= keep) return

        // Sort by filename string (incl. -N suffix) ascending — lexicographically oldest first.
        val sorted = files.sortedBy { it.name }
        val toDelete = sorted.take(files.size - keep)
        for (path in toDelete) {
            fileSystem.delete(path, mustExist = false)
        }
    }

    // ══════════════════════════════════════════════════════════════════════
    // RecoveryRepository — read path
    // ══════════════════════════════════════════════════════════════════════

    /**
     * Returns notes newest-first by parsed [RecoveryInstant].
     *
     * Guards with `exists` before calling `list` (okio `list` throws on an absent dir).
     * Skips filenames that [RecoveryFilename.parse] cannot parse. Decodes at most
     * [RecoveryPreview.PREVIEW_READ_LIMIT] bytes leniently via `Buffer.readUtf8()`.
     * Returns empty list when the dir is absent — never creates it (FR-16, FR-17).
     */
    override fun list(): List<RecoveryNote> {
        if (!fileSystem.exists(recoveryDir)) return emptyList()

        val notes = mutableListOf<RecoveryNote>()
        for (path in listTxtFiles()) {
            val savedAt = RecoveryFilename.parse(path.name) ?: continue // skip malformed
            val head = readHead(path)
            val preview = RecoveryPreview.truncate(head)
            notes += RecoveryNote(path = path.toString(), savedAt = savedAt, preview = preview)
        }

        // Newest-first: descending by savedAt (compare field-by-field via the natural data class order)
        notes.sortWith(compareByDescending<RecoveryNote> { it.savedAt.year }
            .thenByDescending { it.savedAt.month }
            .thenByDescending { it.savedAt.day }
            .thenByDescending { it.savedAt.hour }
            .thenByDescending { it.savedAt.minute }
            .thenByDescending { it.savedAt.second }
            .thenByDescending { it.savedAt.millis })
        return notes
    }

    /**
     * Returns the full UTF-8 text of the file at [path], or `null` if the file has vanished.
     * Never throws on an absent file (FR-18, R-A4).
     */
    override fun read(path: String): String? {
        val okioPath = path.toPath()
        if (!fileSystem.exists(okioPath)) return null
        return try {
            fileSystem.source(okioPath).use { source ->
                val buffer = Buffer()
                source.read(buffer, Long.MAX_VALUE)
                buffer.readUtf8()
            }
        } catch (e: IOException) {
            // File may have vanished between the exists check and the read (tolerate)
            null
        }
    }

    /**
     * Deletes the file at [path]. No-op when the file or directory is absent.
     * Uses `mustExist = false` (FR-18, EC-03).
     */
    override fun delete(path: String) {
        if (!fileSystem.exists(recoveryDir)) return
        fileSystem.delete(path.toPath(), mustExist = false)
    }

    /**
     * Deletes every `.txt` file in [recoveryDir]. No-op when the dir is absent (FR-16).
     */
    override fun deleteAll() {
        if (!fileSystem.exists(recoveryDir)) return
        for (path in listTxtFiles()) {
            fileSystem.delete(path, mustExist = false)
        }
    }

    // ══════════════════════════════════════════════════════════════════════
    // Private helpers
    // ══════════════════════════════════════════════════════════════════════

    /**
     * Lists all `.txt` files under [recoveryDir] WITHOUT sorting (callers sort).
     * Caller must guard with `fileSystem.exists(recoveryDir)` before calling.
     */
    private fun listTxtFiles(): List<Path> =
        fileSystem.list(recoveryDir).filter { it.name.endsWith(".txt") }

    /**
     * Reads at most [RecoveryPreview.PREVIEW_READ_LIMIT] bytes from [path]'s head.
     *
     * Uses `okio.Buffer.readUtf8()` for lenient UTF-8 decoding — equivalent to Dart's
     * `utf8.decode(bytes, allowMalformed: true)`. Does not throw when the window ends
     * mid-UTF-8 sequence; U+FFFD substitution is acceptable (FR-06, NFR-04).
     *
     * Returns an empty string if the file has vanished since listing (tolerate R-06).
     */
    private fun readHead(path: Path): String {
        return try {
            fileSystem.source(path).use { source ->
                val buffer = Buffer()
                source.read(buffer, RecoveryPreview.PREVIEW_READ_LIMIT.toLong())
                buffer.readUtf8()
            }
        } catch (e: IOException) {
            // File vanished between list() and this read — tolerate (R-06, FR-18)
            ""
        }
    }

    /**
     * Resolves the write target for [stem], appending `-1`, `-2`, … on collision.
     * Never overwrites an existing file (FR-13, EC-07).
     */
    private fun resolveFile(stem: String): Path {
        var candidate = recoveryDir / "$stem.txt"
        var suffix = 0
        while (fileSystem.exists(candidate)) {
            suffix += 1
            candidate = recoveryDir / "$stem-$suffix.txt"
        }
        return candidate
    }

    /**
     * Builds the colon-free ms filename stem from the 7 calendar fields of [instant].
     *
     * Format: `YYYY-MM-DDTHH-MM-SS-mmmZ`
     * Pure string assembly — no epoch-millis date math, no kotlinx-datetime (NFR-05).
     * The millisecond component is zero-padded to 3 digits (e.g. `009` not `9`).
     */
    private fun buildStem(instant: RecoveryInstant): String {
        fun pad(value: Int, width: Int) = value.toString().padStart(width, '0')
        return "${pad(instant.year, 4)}-${pad(instant.month, 2)}-${pad(instant.day, 2)}" +
                "T${pad(instant.hour, 2)}-${pad(instant.minute, 2)}-${pad(instant.second, 2)}" +
                "-${pad(instant.millis, 3)}Z"
    }
}
